File: c++/cote/GJM/게임판_덮기2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>


using namespace std;
typedef vector<pair<int,int>> rotation;

int covered[10][10];
int best;


vector<int>& split(const string& s ){
	vector<int>* ret = new vector<int>();

	stringstream ss(s);
	string tmp;

	while(getline(ss,tmp,' ')){
		ret->push_back(stoi(tmp));
	}

	return *ret;
}


vector<string>& get_object(fstream& fd, const int H, const int W ){

	vector<string>* ret = new vector<string>();
	string buf;

	for(int i=0; i< H; i++){
		getline(fd,buf);
		ret->push_back(buf.substr(0,W));
	}


	return *ret;
}


vector<string> rotate(const vector<string>& arr){

	vector<string> ret(arr[0].size(), string(arr.size(), ' '));
	
	for(int i =0; i< arr.size(); i++)
		for(int j=0; j< arr[0].size(); j++)
			ret[j][arr.size()-i-1] = arr[i][j];

	return ret;
}

vector<rotation> generateRotations(vector<string> block){
	vector<rotation> ret (4);

	for(int k= 0; k< 4; k++){
		int originY = -1, originX =-1;
		for(int i=0; i < block.size(); i++){
			for(int j=0; j< block[i].size(); j++){
				if(block[i][j] == '#'){
					if(originY == -1){
						originY = i;
						originX = j;
					}
					ret[k].push_back(make_pair( i- originY, j- originX));
				}
			}
		}
		block = rotate(block);
	}

	sort(begin(ret), end(ret));
	ret.erase(unique(begin(ret), end(ret)), end(ret));

	return ret;
}

bool set(int y, int x, const rotation& r, int xlim, int ylim, int delta){
	bool answer = true;

	for(int i=0; i<r.size(); i++){
		
		int nx = x + r[i].second;
		int ny = y + r[i].first;

		if( nx < 0 || nx >= xlim  || ny <0 || ny >= ylim || (covered[ny][nx]+= delta) > 1){
			answer = false;
		}

	}


	return answer;
}


int simpleHeuristic(int H, int W, int r){
	
	int count = 0;

	for(int i =0; i< H; i++)
		for(int j=0; j<W; j++)
			if(covered[i][j] == 0) count+=1;

	return count / r;
}


void search(const vector<string>& board, const vector<rotation>& rotations, const int H, const int W, int count){

	if ( count + simpleHeuristic(H,W,rotations[0].size()) <  best ) return;

	int y= -1, x= -1;

	for(int i=0; i< H; i++){
		for(int j=0; j<W; j++)
			if(covered[i][j] == 0){
				y=i;
				x=j;
				break;
			}
		if(y != -1) break;
	}

	if( y == -1 ){
		best = max(best, count);
		return;
	}

	for(int i=0; i<rotations.size(); i++){
		if(set(y,x, rotations[i], H, W, 1))
			search(board, rotations, H, W, count +1);
		set(y,x, rotations[i], H, W, -1);
	}

	covered[y][x] = 1;

	search(board, rotations, H ,W, count);


	return ;
} 

int solve(vector<string>& board, vector<string>& block, vector<int>& n){

	vector<rotation> rotations = generateRotations(block);
	best = 0;

	for(int i=0; i<n[0]; i++)
		for(int j =0; j<n[1]; j++)
			covered[i][j] = (board[i][j] == '#' ? 1 : 0);

	search(board, rotations, n[0], n[1], 0);

	return best;
}



int main(void){

	fstream fd("data/게임판_덮기2.txt");
	string buf;

	getline(fd,buf);
	int test_cases = stoi(buf);

	while(test_cases > 0){

		// [H, W, R, C]
		getline(fd, buf);
		vector<int> n = split(buf);

		vector<string> board = get_object(fd, n[0], n[1]);
		vector<string> block = get_object(fd, n[2], n[3]);

		cout << solve(board, block, n)<< endl;
		test_cases--;
	}




	return 0;
}

File: java/src/main/java/com/example/cote/GJM/게임판_덮기2.java
package com.example.cote.GJM;


import java.nio.file.*;
import java.io.*;
import java.util.*;
import java.util.stream.*;

public class 게임판_덮기2 {

	private static int best ;

	private static class Point implements Comparable {
		Integer x;
		Integer y;

		Point(int y, int x){
			this.y = y;
			this.x = x;
		}

		public int compareTo(Object c){
			if(c == this) return 0;
			if(c == null) return -1;
			if(getClass() != c.getClass()) return -1; 

			Point newC = (Point) c;
			if((this.y).compareTo(newC.y) > 0) return 1;
			else return  (this.x).compareTo(newC.x);
		}

		public boolean equals(Object c){
			if(c == this) return true;
			if(c == null) return false;
			if(getClass() != c.getClass()) return false; 

			Point newC = (Point) c;

			if( y != newC.y) return false;
			if (x != newC.x) return false;

			return true;
		}

		public int hashCode(){
			return Objects.hash(x)+Objects.hash(y*10);
		}

	}

	private static int[][] getObject(BufferedReader fd, int H, int W) throws IOException{

		int[][] ret = new int[H][W];

		for(int j =0; j< H; j++){
			String row = fd.readLine();
			for(int i=0; i<W; i++){
				ret[j][i] = (((Character)'#').equals(row.charAt(i))) ? 1 : 0;
			}
		}

		return ret;
	}

	private static int[][] rotate(int[][] block){

		int h = block.length;
		int w = block[0].length;
		int[][] ret = new int[w][h];

		for(int i=0 ; i < h; i++){
			for(int j=0; j< w; j++){
				ret[j][h-1-i] = block[i][j];
			}
		}

		return ret;
	}

	private static	Set<List<Point>> makeRotations(int[][] block, int R, int C){
		
		int x = -1,  y= -1;

		for(int i=0; i< R; i++){
			for(int j=0; j<C; j++){
				if(block[i][j] != 0){
					x = j;
					y = i;
				}
			}
			if(y != -1 ) break;
		}

		Set<List<Point>> ret = new HashSet<>();

		for(int k=0; k<4; k++){
			
			List<Point> r = new LinkedList<Point>();

			for(int i=0; i<block.length; i++){
				for(int j=0; j<block[0].length; j++){
					
					if(block[i][j] > 0 ) 
						r.add(new Point(i-y,j-x));
				}
			}

			ret.add(r);
			Collections.sort(r);
			block = rotate(block);
		}

		return ret;
	}

	private static boolean set(int y, int x ,List<Point> rotation, int[][] board, int delta){
		boolean ret = true;
		int H = board.length;
		int W = board[0].length;
		
		for(Point p : rotation){
			int nx = x + p.x;
			int ny = y + p.y;

			if( nx < 0 || nx >= W || ny < 0 || ny >= H || (board[ny][nx] += delta) > 1){
				ret = false;
			}
		}

		return ret;
	}

	private static int simpleHeuristic(int[][] board, int r){
		int count = 0;

		for(int i=0; i<board.length; i++){
			for(int j=0; j< board[0].length; j++){
				if(board[i][j] == 0) count++;
			}
		}

		return count/ r;
	}

	private static void search(int[][] board, Set<List<Point>> rotations, int count){

		if( count + simpleHeuristic(board, rotations.iterator().next().size()) < best) return;

		int x= -1, y= -1;

		for(int i= 0; i<board.length; i++){
			for(int j=0; j<board[0].length; j++){
				if(board[i][j] == 0 ){
					y = i;
					x = j;
					break;
				}
			}
			if(y!= -1) break;
		}

		int H = board.length;
		int W = board[0].length;

		if( y == -1 ){
			best = Math.max(best, count);
			return;
		}

		for(List<Point> rotation : rotations){
			if(set(y, x, rotation, board, 1))
				search(board,rotations, count+1);
			set(y,x,rotation,board,-1);
		}
		
		board[y][x] = 1;

		search(board, rotations, count);

		return;

	}


	private static int solve(int[][] board,  int[][] block, int H, int W, int R, int C){

		Set<List<Point>> rotations = makeRotations(block, R, C);
		best = 0;
		search(board, rotations, 0 );
		return best;
	}

	public static void main(String... args) throws IOException{

		Path p = Paths.get(System.getProperty("user.dir")+"/data/게임판_덮기2.txt");
		BufferedReader fd = Files.newBufferedReader(p);

		int test_cases = Integer.parseInt(fd.readLine());

		while( test_cases > 0 ){
			List<Integer> n = Arrays.stream(fd.readLine().split(" "))
									.map(Integer::parseInt)
									.collect(Collectors.toList());

			int[][] board = getObject(fd, n.get(0), n.get(1));
			int[][] block = getObject(fd, n.get(2), n.get(3));

			System.out.println(solve(board, block, n.get(0), n.get(1), n.get(2), n.get(3)));

			test_cases--;

		}



	}

}